soroban-env-common/src/env.rs

```diff
@@ -461,6 +467,23 @@ macro_rules! call_macro_with_all_host_functions {
                 {"0", fn verify_sig_ed25519(x:Object, k:Object, s:Object) -> RawVal}
             }
 
+            mod prng "p" {
+                /// Return a u32 from the persistent, stateful PRNG associated
+                /// with the calling context, finalizing the PRNG if `finalize`
+                /// is set to `true`. Traps if PRNG is already finalized.
+                {"_", fn prng_next_u32(finalize:RawVal) -> RawVal}
+                /// Return a u63 from the persistent, stateful PRNG associated
+                /// with the calling context, finalizing the PRNG if `finalize`
+                /// is set to `true`. Traps if PRNG is already finalized.
+                {"0", fn prng_next_u63(finalize:RawVal) -> RawVal}
+                /// Fill the region of linear memory between `lm_pos` and
+                /// `lm_pos` + `len` with bytes drawn from from the persistent,
+                /// stateful PRNG associated with the calling context,
+                /// finalizing the PRNG if `finalize` is set to `true`. Traps if
+                /// PRNG is already finalized. Returns void.
+                {"1", fn prng_fill_linear_memory(lm_pos:RawVal, len:RawVal, finalize:RawVal) -> RawVal}
```

Comment:
💭 In the SDK I can see us implementing an interface similar to that of the [`rand::Rng`](https://docs.rs/rand/latest/rand/trait.Rng.html) type, although much simpler, something in the form:

```
pub trait ... {
    fn gen<T>(&mut self) -> T;
    fn fill<T>(&mut self, dest: &mut T);
}
```

Where `gen` would probably support primitive types that we support across boundaries, like `u32`, `u64`, `i32`, `i64`, and maybe smaller types too I guess like 8-bit, 16-bit, and the size integer types.

Where `fill` would probably support byte slices, byte arrays, and `Bytes` and `BytesN` types.

For the `gen` case:
- ❓ I'd like `gen` to only return positive values for the integer types. That's what the Go stdlib's math/rand and crypto/rand functions do, and in my experience that is 99% what users want. However, it seems like the behavior of `rand::Rng` is to return values in the full range of the type. It might be better to do the full range way for consistency with other Rust programs. Wdyt @graydon?
- ✅  `prng_next_u32` looks good and we'd use it for the 8-bit, 16-bit, and 32-bit cases, truncating it in the formers.
- ❓ `prng_next_u63` looks good if we only support positive values on types like i64 if we model the Go API, but could be difficult to use if we model the Rust rand::Rng API, since in the SDK we'll surface it as a `i64` but in practice it will only ever generate half the range. We could add a special function like `gen_pos_i64` specifically, that would call through to this, that'd work.
- ❓ If we want full range integer types, what would be the recommendation for getting a u64, or full range i64? We probably need a `prng_next_u64` that we can use for those two cases.

For the `fill` case:
- ✅ The API looks good.

Reply:
Re: positive values, and the API shape in general: I think if we say we're generating a type, we should return the full range of the type.

But thinking about it a bit more, I also think we might want to steer users towards slightly higher-level APIs, which are what they usually want. Random number generation doesn't happen in a vacuum, it's usually task-driven. We could do some of these in the host also, they're standard-enough (and use host objects enough) to do there:

- Shuffle a host vec (do in host)
- Select a random element of a host vec or a map (do in host)
- Generate a value in a given `[lo,hi]` value-range (maybe do in guest)
- Generate a value with a given symbolic distribution (maybe do in guest)

I could definitely modify it to return i64 or u64 full values, if thoroughness-wrt-number-types is of interest, I just figured I'd do a non-object-allocating interface. Not sure which scenarios users are going to really want/need (or what our plans are for reforming the number types in general).

Reply:
All of those things sound good to me.

The last two dot points I think the host should provide too. We may as well?

For i64 / u64 would it be okay for me to just make 2x next_u32 calls to construct a hilo? If that's the case, that's probably more efficient than doing one host fn call that constructs a host object which then has to be unpeeled via an extra host function call anyway.

For the select random element, that's essentially the same thing generating a value in a range, if we add the ability to access a map with an index.

Reply:
I would stay away for now (as host functions) from providing those extra things as it's unlikely to be perf critical: the more useful thing to do is to be sure that we provide the right low level building blocks that will be implemented correctly.

At a minimum:
* fill an `[u8]` with random data (you already have that)
* construct "types of interest" from `[u8]`
   * I see that we already support this for `BigNum`. For now this is probably enough for "all integers" (including i64) as `BigNum` has `to_i64`
   * Should it be possible to generate random symbols, strings too?

Reply:
It's common in the Rust stdlib + community `rand` crate to provide a `Vec::choose` method, which can be given an RNG to choose a random value from the Vec.

I think what @graydon is suggesting here by offering host functions to select a random element from a Soroban Vec is approximately equivalent.

We can definitely do that on the SDK side instead, but it also seems like something the host could do entirely.

Reply:
I'm less concerned with the performance benefit of doing it in the host at this point, and more interested in the fact that if we do that, we keep as much rand logic as possible in the host, so the SDK can be rather unintelligent about its provision of random capabilities.

Reply:
I don't think we know what people will actually want to do with those things, and the more we add (in the SDK or in the host) the more we have to specify and test. It seems easy to add in the host or in SDKs later as part of the normal process.

Reply:
> We can punt on container-select / leave it to the guest if we're willing to expose a map-get-by-numeric-index method. We don't have one currently but it's plausible and more general, I don't mind adding that.

I know I suggested the indexing idea for maps a few messages up, but I regret is and want to walk it back. Reason being is that reflecting on it, this would probably be the only place where we'd use the index. And an index into a map would be in many ways a footgun to expose in the SDK so we'd probably just use it internally. So we'd be adding this index to access a random element, but not using the index elsewhere, so we may as well just have a host function that returns a random key from the map. Which I think is what you originally suggested.

Reply:
I think adding this without having a way to transform vectors is a missed opportunity to make it better for other real use cases. Spinning up a thread on Discord...

Reply:
`swap` might be nice to add, but isn't adequate -- it still produces a new vector on each call. immutable objects! the point is to enable creating a shuffled vector efficiently, without O(n) intermediates.

I agree a number-in-a-range API is probably one of the more useful things to provide, to avoid footguns.